p2p/transport/quic/virtuallistener.go

```diff
@@ -0,0 +1,168 @@
+package libp2pquic
+
+import (
+	"errors"
+	"sync"
+
+	tpt "github.com/libp2p/go-libp2p/core/transport"
+	"github.com/libp2p/go-libp2p/p2p/transport/quicreuse"
+	"github.com/lucas-clemente/quic-go"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+const acceptBufferPerVersion = 4
+
+// virtualListener is a listener that exposes a single multiaddr but uses another listener under the hood
+type virtualListener struct {
+	*listener
+	udpAddr       string
+	version       quic.VersionNumber
+	t             *transport
+	acceptRunnner *acceptLoopRunner
+	acceptChan    chan acceptVal
+}
+
+var _ tpt.Listener = &virtualListener{}
+
+func (l *virtualListener) Multiaddr() ma.Multiaddr {
+	return l.listener.localMultiaddrs[l.version]
+}
+
+func (l *virtualListener) Close() error {
+	l.acceptRunnner.rmAcceptForVersion(l.version)
+
+	l.t.listenersMu.Lock()
+	defer l.t.listenersMu.Unlock()
+
+	var err error
+	listeners := l.t.listeners[l.udpAddr]
+	if len(listeners) == 1 {
+		// This is the last virtual listener here, so we can close the underlying listener
+		err = l.listener.Close()
+		delete(l.t.listeners, l.udpAddr)
+	} else {
+		for i := 0; i < len(listeners); i++ {
+			// Swap remove
+			if l == listeners[i] {
+				listeners[i] = listeners[len(listeners)-1]
+				listeners = listeners[0 : len(listeners)-1]
+				l.t.listeners[l.udpAddr] = listeners
+				break
+			}
+		}
+	}
+
+	return err
+}
+
+func (l *virtualListener) Accept() (tpt.CapableConn, error) {
+	var v acceptVal
+	var ok bool
+	select {
+	// Check if we have a pending connection first
+	case v, ok = <-l.acceptChan:
+	default:
+		// No? Let's call Accept and wait for a connection
+		go l.acceptRunnner.Accept(l.listener, l.version)
```

Comment:
Not sure I understand this logic here, especially the pending first connection.

I was wondering if `Accept` could this be driven by the Go routines that called `Accept`? It would be really nice if we didn't have to start another Go routine here. 
You could then retrieve connections from the underlying listener, and return them directly (if they're intended for this virtual listener), or enqueue them (if they're intended for a virtual listener of a different version).

Reply:
Refactored to get rid of the goroutine. We still check first if we have previously enqueued a connection.